src/common/cockpitconf.c

```diff
@@ -21,282 +21,393 @@
-          guint i;
-          value = g_new (const gchar *, n + 1);
-
-          for (i = 0; i < n; i++)
-            value[i] = g_strdup (strv[i]);
-
-          value[i] = NULL;
-          g_strfreev (strv);
-        }
-    }
-  g_free (stripped);
-  return value;
+  const char *value = cockpit_conf_string (section, field);
+  if (value)
+    return strcasecmp (value, "yes") == 0 || strcasecmp (value, "true") == 0 || strcmp (value, "1") == 0;
+  return defawlt;
 }
 
-const gchar **
-cockpit_conf_strv (const gchar *section,
-                   const gchar *field,
-                   gchar delimiter)
+const char **
+cockpit_conf_strv (const char *section,
+                   const char *field,
+                   char delimiter)
 {
-  const gchar **value = NULL;
-  gchar *key = NULL;
-  gchar delm[2] = {delimiter, '\0'};
-
-  g_return_val_if_fail (section != NULL, NULL);
-  g_return_val_if_fail (field != NULL, NULL);
+  Entry *entry = cockpit_conf_lookup (section, field);
 
-  ensure_cockpit_conf ();
-  key = g_strjoin (delm, section, field, NULL);
+  if (!entry || !entry->value)
+    return NULL;
 
-  if (!g_hash_table_contains (cached_strvs, key))
+  if (entry->strv_cache)
     {
-      value = build_strv (section, field, delm);
-      if (value)
-        g_hash_table_insert (cached_strvs, g_strdup (key), value);
+      if (delimiter != entry->strv_delimiter)
+        errx (1, "cockpitconf: Looking up strv with different delimiters is not supported");
     }
   else
     {
-      value = g_hash_table_lookup (cached_strvs, key);
+      /* strip off trailing whitespace (leading whitespace is already stripped by regexp) */
+      entry->strv_value = strdupx (entry->value);
+      for (char *c = entry->strv_value + strlen (entry->strv_value) - 1; c >= entry->strv_value && isspace (*c); --c)
+        *c = '\0';
+      entry->strv_cache = strsplit (entry->strv_value, delimiter);
+      entry->strv_delimiter = delimiter;
     }
 
-  g_free (key);
-  return value;
+  return entry->strv_cache;
 }
 
-guint
-cockpit_conf_guint (const gchar *section,
-                    const gchar *field,
-                    guint default_value,
-                    guint64 max,
-                    guint64 min)
+unsigned
+cockpit_conf_uint (const char *section,
+                   const char *field,
+                   unsigned default_value,
+                   unsigned max,
+                   unsigned min)
 {
-  guint val = default_value;
-  guint64 conf_val;
-  gchar *endptr = NULL;
+  unsigned val = default_value;
+  long long conf_val;
+  char *endptr = NULL;
 
-  const gchar* conf = cockpit_conf_string (section, field);
+  const char* conf = cockpit_conf_string (section, field);
   if (conf)
     {
-      conf_val = g_ascii_strtoull (conf, &endptr, 10);
-      if ((conf_val == G_MAXUINT64 || conf_val == 0) &&
+      errno = 0;
+      conf_val = strtoll (conf, &endptr, 10);
+      if ((conf_val == LLONG_MAX || conf_val == 0) &&
```

Comment:
This is signed now, so you want to check for `LLONG_MIN`.

Reply:
Fixed, thanks for spotting!